### DISCUSSION TITLE
Seeking Help with Implementing AvaloniaUI Interfaces for Windowing

### DISCUSSION THREAD

**Discussion by DenisMontes** [AvaloniaUI/Avalonia#12654] [Q&A] [answered]
Dear AvaloniaUI Community,

I hope this message finds you well. I am writing to seek your assistance regarding an issue I've encountered while attempting to integrate AvaloniaUI with a 3D rendering library, such as OpenTK or Silk.NET, to create windows for rendering 3D scenes.

From my understanding, in order to allow AvaloniaUI to compose its controls on top of these rendering windows, I would need to implement the IWindowingPlatform interface to create the windows and the IWindowImpl interface within the windows to make them recognizable as valid Avalonia windows. However, as I've been attempting to implement these interfaces (using Avalonia version 11.0.4), I am encountering an error message stating that I cannot implicitly implement an inaccessible member.

I have reviewed the documentation and code examples available, and I believe I am following the correct approach. However, this error has left me puzzled, and I'm wondering if there might be some detail that I've overlooked or misunderstood in the process.

Has anyone else in the community faced a similar issue or encountered this particular error message while implementing the IWindowingPlatform and IWindowImpl interfaces? If so, could you kindly provide some guidance on how to proceed or if there's a workaround to resolve this error? I would greatly appreciate any insights, tips, or suggestions that you can provide to help me overcome this obstacle.

Thank you in advance for your time and assistance. I value the vibrant knowledge-sharing environment of the AvaloniaUI community and am hopeful that together we can find a solution to this challenge.

**Comment by stevemonaco** [comment]
> I would need to implement the IWindowingPlatform interface to create the windows and the IWindowImpl interface within the windows to make them recognizable as valid Avalonia windows. However, as I've been attempting to implement these interfaces (using Avalonia version 11.0.4), I am encountering an error message stating that I cannot implicitly implement an inaccessible member.

APIs marked with `[PrivateApi]` aren't for "normal" Avalonia use. They are made internal for the reference assemblies during Avalonia's build process. I believe you need to consume the runtime assemblies instead. [Estragonia](https://github.com/MrJul/Estragonia/blob/main/src/JLeb.Estragonia/JLeb.Estragonia.csproj) has an example of how to do this while implementing Avalonia support for within Godot. I'm not sure if this is the "official" way to support new platforms.

I would follow Estragonia's approach as you probably want to render into a texture instead. You will probably reuse the windowing capabilities provided from GLFW or possibly not support windowing within Avalonia. It does mean you'll need to adapt lots of input-related and other APIs that Avalonia needs, too.

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/wiki/Using-private-apis-in-nuget-packages

**Comment by DenisMontes** [comment, marked as answer]
Thank you. I found this answer:  https://github.com/AvaloniaUI/Avalonia/discussions/12648#discussioncomment-6821275. It works for me.
